firedrake/mg/ufl_utils.py

```diff
@@ -156,41 +172,58 @@ def coarsen_function_space(V, self, coefficient_mapping=None):
     except ValueError:
         # Not in an add_hooks context
         pass
-
     return V
 
 
+@coarsen.register(firedrake.Cofunction)
 @coarsen.register(firedrake.Function)
 def coarsen_function(expr, self, coefficient_mapping=None):
     if coefficient_mapping is None:
         coefficient_mapping = {}
     new = coefficient_mapping.get(expr)
     if new is None:
-        V = expr.function_space()
-        manager = firedrake.dmhooks.get_transfer_manager(expr.function_space().dm)
-        V = self(expr.function_space(), self)
-        new = firedrake.Function(V, name="coarse_%s" % expr.name())
-        manager.inject(expr, new)
-    return new
-
-
-@coarsen.register(firedrake.Constant)
-def coarsen_constant(expr, self, coefficient_mapping=None):
-    if coefficient_mapping is None:
-        coefficient_mapping = {}
-    new = coefficient_mapping.get(expr)
-    if new is None:
-        mesh = self(extract_unique_domain(expr), self)
-        new = firedrake.Constant(numpy.zeros(expr.ufl_shape,
-                                             dtype=expr.dat.dtype),
-                                 domain=mesh)
-        # Share data pointer
-        new.dat = expr.dat
+        Vf = expr.function_space()
+        Vc = self(Vf, self)
+        new = firedrake.Function(Vc, name="coarse_%s" % expr.name())
```

Comment:
Remind me, `firedrake.Function` makes a `CoFunction` if `Vc` is a dual space?

Reply:
Correct